servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/FromNPublisher.java

```diff
@@ -17,47 +17,58 @@
 
 import javax.annotation.Nullable;
 
+import static io.servicetalk.concurrent.internal.FlowControlUtils.addWithOverflowProtection;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.handleExceptionFromOnSubscribe;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
 import static io.servicetalk.concurrent.internal.SubscriberUtils.newExceptionForInvalidRequestN;
+import static java.lang.Math.min;
 
-final class From2Publisher<T> extends AbstractSynchronousPublisher<T> {
+final class FromNPublisher<T> extends AbstractSynchronousPublisher<T> {
     @Nullable
     private final T v1;
     @Nullable
     private final T v2;
+    @Nullable
+    private final T v3;
```

Comment:
The two independent classes were introduced to reduce memory consumption as they are commonly used cases. After the classes are combined are there any benefit relative to `FromArrayPublisher` anymore? Looks like the `From2` will use more memory. Can you fix the issues in the same code structure and explore combining memory impact as followup PR?

Reply:
@Scottmitch I did some changes, share your thoughts. 

**Before** 

```
io.servicetalk.concurrent.api.From2Publisher object internals:
OFF  SZ   TYPE DESCRIPTION               VALUE
  0   8        (object header: mark)     N/A
  8   4        (object header: class)    N/A
 12   4      T From2Publisher.v1         N/A
 16   4      T From2Publisher.v2         N/A
 20   4        (object alignment gap)    
Instance size: 24 bytes
Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
```

```
io.servicetalk.concurrent.api.From2Publisher.TwoValueSubscription object internals:
OFF  SZ                    TYPE DESCRIPTION                       VALUE
  0   8                         (object header: mark)             N/A
  8   4                         (object header: class)            N/A
 12   1                    byte TwoValueSubscription.state        N/A
 13   3                         (alignment/padding gap)           
 16   4   Subscriber<? super T> TwoValueSubscription.subscriber   N/A
 20   4                         (object alignment gap)            
Instance size: 24 bytes
Space losses: 3 bytes internal + 4 bytes external = 7 bytes total
```

------

```
io.servicetalk.concurrent.api.From3Publisher object internals:
OFF  SZ   TYPE DESCRIPTION               VALUE
  0   8        (object header: mark)     N/A
  8   4        (object header: class)    N/A
 12   4      T From3Publisher.v1         N/A
 16   4      T From3Publisher.v2         N/A
 20   4      T From3Publisher.v3         N/A
Instance size: 24 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

```
io.servicetalk.concurrent.api.From3Publisher.ThreeValueSubscription object internals:
OFF  SZ                    TYPE DESCRIPTION                         VALUE
  0   8                         (object header: mark)               N/A
  8   4                         (object header: class)              N/A
 12   1                    byte ThreeValueSubscription.state        N/A
 13   3                         (alignment/padding gap)             
 16   4   Subscriber<? super T> ThreeValueSubscription.subscriber   N/A
 20   4                         (object alignment gap)              
Instance size: 24 bytes
Space losses: 3 bytes internal + 4 bytes external = 7 bytes total
```

-----

```
io.servicetalk.concurrent.api.FromArrayPublisher object internals:
OFF  SZ   TYPE DESCRIPTION                 VALUE
  0   8        (object header: mark)       N/A
  8   4        (object header: class)      N/A
 12   4    T[] FromArrayPublisher.values   N/A
Instance size: 16 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

```
io.servicetalk.concurrent.api.FromArrayPublisher.FromArraySubscription object internals:
OFF  SZ                    TYPE DESCRIPTION                            VALUE
  0   8                         (object header: mark)                  N/A
  8   4                         (object header: class)                 N/A
 12   4                     int FromArraySubscription.beginOffset      N/A
 16   4                     int FromArraySubscription.endOffset        N/A
 20   1                 boolean FromArraySubscription.ignoreRequests   N/A
 21   3                         (alignment/padding gap)                
 24   4   Subscriber<? super T> FromArraySubscription.subscriber       N/A
 28   4                     T[] FromArraySubscription.values           N/A
Instance size: 32 bytes
Space losses: 3 bytes internal + 0 bytes external = 3 bytes total
```
-----

**After**

```
io.servicetalk.concurrent.api.FromNPublisher object internals:
OFF  SZ   TYPE DESCRIPTION               VALUE
  0   8        (object header: mark)     N/A
  8   4        (object header: class)    N/A
 12   4      T FromNPublisher.v1         N/A
 16   4      T FromNPublisher.v2         N/A
 20   4      T FromNPublisher.v3         N/A
Instance size: 24 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

```
io.servicetalk.concurrent.api.FromNPublisher object internals:
OFF  SZ   TYPE DESCRIPTION               VALUE
  0   8        (object header: mark)     N/A
  8   4        (object header: class)    N/A
 12   4      T FromNPublisher.v1         N/A
 16   4      T FromNPublisher.v2         N/A
 20   4      T FromNPublisher.v3         N/A
Instance size: 24 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

----

With that in mind, before & after have no difference in heap cost between them. The new version still saves 8bytes per subscription instance, but if to fwiw it doesn't justify the loss of code readability. Last, the n-arity operator performs better against the array version when it comes to allocation rate.

Reply:
thanks for providing this analysis. as long as no memory increase lgtm!